pkg/river/vm/vm.go

```diff
@@ -85,16 +85,47 @@ func (vm *Evaluator) evaluateBlockOrBody(scope *Scope, assoc map[value.Value]ast
 		rv = rv.Addr()
 	}
 
+	return vm.evaluateBlockOrBodyDefault(scope, assoc, node, rv)
+}
+
+func (vm *Evaluator) evaluateBlockOrBodyDefault(scope *Scope, assoc map[value.Value]ast.Node, node ast.Node, rv reflect.Value) error {
+	if ru, ok := rv.Interface().(value.Defaulter); ok {
+		ru.SetToDefault()
+	}
+
+	return vm.evaluateBlockOrBodyUnmarshal(scope, assoc, node, rv)
+}
+
+func (vm *Evaluator) evaluateBlockOrBodyUnmarshal(scope *Scope, assoc map[value.Value]ast.Node, node ast.Node, rv reflect.Value) error {
```

Comment:
I'm not sure about the order here. It feels like Unmarshal should take precedence over Defaults, so that it becomes:

1. Unmarshaler 
2. Defaulter
3. (decode)
4. Validator 

This is because Unmarshaler primarily allows you to change the type used for decoding, which we do constantly in all of our implementations. It feels like if you're changing the type, that new type should be checked for Defaulter/Validator instead.

WDYT?

Reply:
That seems reasonable and I'll experiment with this order.